controllers/tykoasapidefinition_controller.go

```diff
@@ -196,6 +232,182 @@ func (r *TykOasApiDefinitionReconciler) Reconcile(ctx context.Context, req ctrl.
+			Namespace: versionedTykOAS.Status.GetBaseVersionNamespace(),
+		}, &baseOasAPI); err != nil {
+			if k8serrors.IsNotFound(err) {
+				log.Info("failed to get TykOasApiDefinition CR", "err", err.Error())
+			}
+
+			return err
+		}
+
+		errMsg := "Cannot delete API. It is referenced as version in another Tyk OAS API. Please remove it before deleting."
+
+		versionedTykOAS.Status.LatestTransaction = v1alpha1.TransactionInfo{
+			Time:   metav1.Now(),
+			Status: v1alpha1.TransactionStatus(errMsg),
+		}
+
+		log.Info("error deleting versioned TykOasAPI", errMsg)
+
+		if err := r.Status().Update(ctx, versionedTykOAS); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (r *TykOasApiDefinitionReconciler) cleanUpDanglingVersionedOasAPIs(
+	ctx context.Context,
+	versionedTykOAS *v1alpha1.TykOasApiDefinition,
+	log logr.Logger,
+) error {
+	if versionedTykOAS.Status.GetBaseVersionName() == "" {
+		return nil
+	}
+
+	var baseOasAPI v1alpha1.TykOasApiDefinition
+
+	if err := r.Get(ctx, types.NamespacedName{
+		Name:      versionedTykOAS.Status.GetBaseVersionName(),
+		Namespace: versionedTykOAS.Status.GetBaseVersionNamespace(),
+	}, &baseOasAPI); err != nil {
+		if k8serrors.IsNotFound(err) {
+			log.Info("base oas version not found, detaching version from the base OAS API.")
+
+			versionedTykOAS.RemoveOASVersionStatus()
+
+			if err := r.Status().Update(ctx, versionedTykOAS); err != nil {
+				return err
+			}
+
+			return nil
+		}
+
+		return err
+	}
+
+	var oasAPIPresent bool
+
+	for _, v := range baseOasAPI.Spec.Versioning.Versions {
+		if v.TykOasApiDefinitionRef == versionedTykOAS.Name {
+			oasAPIPresent = true
+			break
+		}
+	}
+
+	if !oasAPIPresent {
+		versionedTykOAS.RemoveOASVersionStatus()
+
+		if err := r.Status().Update(ctx, versionedTykOAS); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// handleVersioningTykOASAPI versions a Base TykOasAPI using refrenced TykOasAPIs.
+func (r *TykOasApiDefinitionReconciler) handleVersioningTykOASAPI(
+	ctx context.Context,
+	tykBaseOASCrd *v1alpha1.TykOasApiDefinition,
+	tykBaseOASDoc *string,
+) error {
+	var versions []*TykOasApiDefinitionVersion
+
+	for _, v := range tykBaseOASCrd.Spec.Versioning.Versions {
+		var versionOasAPI v1alpha1.TykOasApiDefinition
+
+		if err := r.Get(ctx, types.NamespacedName{
+			Name:      v.TykOasApiDefinitionRef,
+			Namespace: v.Namespace,
+		}, &versionOasAPI); err != nil {
+			return err
+		}
+
+		if versionOasAPI.Status.VersioningStatus == nil {
+			versionOasAPI.Status.NewVersioningStatus()
+		}
+
+		if tykBaseOASCrd.Spec.Versioning.Default == v.Name {
+			versionOasAPI.Status.VersioningStatus.IsDefaultVersion = true
+		}
+
+		versionOasAPI.Status.SetBaseVersionName(tykBaseOASCrd.Name)
+		versionOasAPI.Status.SetBaseVersionNamespace(&tykBaseOASCrd.Namespace)
+		versionOasAPI.Status.SetIsVersionedAPI(true)
+
+		version := TykOasApiDefinitionVersion{
+			ID:   versionOasAPI.Status.ID,
```

Comment:
What if versioned API is not created yet and `ID` is empty?

Reply:
If it isn't created, it won't get to there. It will get stop at this [line](https://github.com/TykTechnologies/tyk-operator/pull/719/files#diff-6cd635972dc870184cc2eb7fd04f8ccf46d6821cbec2f194fb419469cc4ba437R359)

Reply:
Can you share line again? It's pointing to empty line right now.

Reply:
[here](https://github.com/TykTechnologies/tyk-operator/pull/719/files#diff-6cd635972dc870184cc2eb7fd04f8ccf46d6821cbec2f194fb419469cc4ba437R357)

Reply:
ID field in status is only set when API is created on Tyk.
So CRD may exists but it might not be created on Tyk yet.